test/playwright/localTest.ts

```diff
@@ -0,0 +1,144 @@
+import * as path from 'path';
+import * as readline from 'readline';
+import * as fs from 'fs/promises';
+import childProcess from 'child_process';
+import { Readable } from 'stream';
+import { once } from 'events';
+import { test as base } from './test';
+
+interface RunningLocalApp {
+  url: string;
+  dir: string;
+}
+
+// You'll need to have `yarn dev13` running for this
+const VERBOSE = true;
+
+async function waitForMatch(input: Readable, regex: RegExp): Promise<RegExpExecArray | null> {
+  return new Promise((resolve, reject) => {
+    const rl = readline.createInterface({ input });
+
+    rl.on('line', (line) => {
+      const match = regex.exec(line);
+      if (match) {
+        rl.close();
+        resolve(match);
+      }
+    });
+    rl.on('error', (err) => reject(err));
+    rl.on('end', () => resolve(null));
+  });
+}
+
+interface WithAppOptions {
+  // Command to start toolpad with
+  cmd?: 'start' | 'dev';
+  // Template to be used as the starting point of the project folder toolpad is running in
+  // This will copied to the temporary folder
+  template?: string;
+  // Run toolpad next.js app in local dev mode
+  toolpadDev?: boolean;
+}
+
+/**
+ * Spins up a Toolpad app in a local temporary folder. The folder is deleted after `doWork` is done
+ */
+export async function withApp(
+  options: WithAppOptions,
+  doWork: (app: RunningLocalApp) => Promise<void>,
+) {
+  const { cmd = 'start', template } = options;
+
+  const projectDir = await fs.mkdtemp(path.resolve(__dirname, './tmp-'));
+
+  try {
+    if (template) {
+      await fs.cp(template, projectDir, { recursive: true });
+    }
+
+    const args: string[] = [cmd];
+    if (options.toolpadDev) {
+      args.push('--dev');
+    }
+
+    const child = childProcess.spawn('toolpad', args, {
+      cwd: projectDir,
+      stdio: 'pipe',
+    });
+
+    try {
+      await Promise.race([
+        new Promise((resolve, reject) => {
```

Comment:
Why is that bad? You should avoid promise constructor as much as possible, I only use it to wrap the child process `exit` event. After I converted it to a promise I just use the regular promise constructs to compose the async behaviors together. The other part of the `race` already has promises, so no need to use `new Promise`.

composing async behavior is much more straightforward with promises than with event handlers and callbacks. That's why it's best to convert events and callbacks to promises asap, and then compose them together with `Promise.all`, `Promise.race`, `async`, `await`

Reply:
Nevermind my bad of course the callback can't be awaited so that's why we need the Promise constructor... I mistakenly thought a function could've worked there. Anyway this was just for consistency, in this case this is fine of course! Sorry for misreading.

Reply:
Actually, Node.js has a `once` utility that we can use.